src/sync/ms_queue.rs

```diff
@@ -185,66 +212,83 @@ impl<T> MsQueue<T> {
-    // Attempt to pop a data node. `Ok(None)` if queue is empty or in blocking
-    // mode; `Err(())` if lost race to pop.
-    fn pop_internal(&self, guard: &epoch::Guard) -> Result<Option<T>, ()> {
-        let head = self.head.load(Acquire, guard).unwrap();
-        if let Some(next) = head.next.load(Acquire, guard) {
+    fn dequeue_internal(&self, guard: &epoch::Guard) -> Result<Option<T>, ()> {
+        // Load the head (with the provideed guard).
+        let head = self.head.load(atomic::Ordering::Acquire, guard).unwrap();
+        // Check if the head is empty or not.
+        if let Some(next) = head.next.load(atomic::Ordering::Acquire, guard) {
+            // The head is not empty; check the payload.
             if let Payload::Data(ref t) = next.payload {
+                // There was some data in the payload, attempt to remove the head.
                 unsafe {
-                    if self.head.compare_and_set_shared(Some(head), Some(next), Release) {
+                    if self.head.compare_and_set_shared(Some(head), Some(next), atomic::Ordering::Release) {
+                        // Unlink the head from the epoch so it can eventually be deallocated.
+                        // TODO: Is this really necessary?
                         guard.unlinked(head);
+
+                        // As we have exclusive accesss (by the CAS above), we can safely move out
+                        // of the pointer.
                         Ok(Some(ptr::read(t)))
                     } else {
+                        // The head has been dequeued by another thread. We lost the race.
                         Err(())
                     }
                 }
             } else {
+                // Nothing in the node.
                 Ok(None)
             }
         } else {
+            // Yep. Empty.
             Ok(None)
         }
     }
 
-    /// Check if this queue is empty.
+    /// Is this queue empty?
     pub fn is_empty(&self) -> bool {
         let guard = epoch::pin();
-        let head = self.head.load(Acquire, &guard).unwrap();
+        let head = self.head.load(atomic::Ordering::Acquire, &guard).unwrap();
 
-        if let Some(next) = head.next.load(Acquire, &guard) {
-            if let Payload::Data(_) = next.payload {
-                false
-            } else {
-                true
-            }
+        if let Some(next) = head.next.load(atomic::Ordering::Acquire, &guard) {
+            // Check if the payload is empty or not.
+            if let Payload::Data(_) = next.payload { false } else { true }
         } else {
+            // There were no node after mandatory head node, so the queue is empty.
             true
         }
     }
 
     /// Attempt to dequeue from the front.
     ///
-    /// Returns `None` if the queue is observed to be empty.
-    pub fn try_pop(&self) -> Option<T> {
+    /// This returns `None` if the queue is observed to be empty.
+    pub fn dequeue(&self) -> Option<T> {
```

Comment:
Rust uses `push_front` and `push_back`, which is only appropriate when you can push/pop from both ends.